src/Core/WinSWCore/ServiceDescriptor.cs

```diff
@@ -95,6 +105,30 @@ public ServiceDescriptor(XmlDocument dom)
             this.dom = dom;
         }
 
+        private void ValidateXmlSchema()
+        {
+            XmlReaderSettings booksettings = new XmlReaderSettings();
+            booksettings.Schemas.Add(null, BasePath + ".xsd");
+            booksettings.ValidationType = ValidationType.Schema;
+            booksettings.ValidationEventHandler += new ValidationEventHandler(booksSettingsValidationEventHandler);
+
+            XmlReader books = XmlReader.Create(BasePath + ".xml");
+
+            while (books.Read()) { }

```

Comment:
Now we read the file twice: once here and once in the constructor. Try to reduce the number of reads to one.

Reply:
Reading and validation combined